src/common/base/src/base/runtime_tracker.rs

```diff
@@ -200,103 +277,109 @@ impl MemoryTracker {
         self.limit.store(size, Ordering::Relaxed);
     }
 
-    /// Accumulate memory usage and check if it exceeds the limit.
+    /// Feed memory usage stat to MemoryTracker and return if it exceeds the limit.
+    ///
+    /// It feeds `state` to the this tracker and all of its ancestors, including GLOBAL_TRACKER.
     #[inline]
-    pub fn record_memory(&self, state: &StatBuffer) {
-        if PANICKING.load(Ordering::Relaxed) {
-            // This thread already panicked, do not panic again, do not push event to parent tracker either.
-            return;
-        }
+    pub fn record_memory(
+        tracker: &Option<Arc<MemoryTracker>>,
+        state: StatBuffer,
+    ) -> Result<(), OutOfLimit> {
+        let mut is_root = false;
+
+        let mem_tracker = match tracker {
+            Some(x) => x,
+            None => {
+                // No parent, report to GLOBAL_TRACKER
+                is_root = true;
+                &GLOBAL_TRACKER
+            }
+        };
 
-        let mut used = self
+        let mut used = mem_tracker
             .memory_usage
             .fetch_add(state.memory_usage, Ordering::Relaxed);
 
         used += state.memory_usage;
 
-        let limit = self.limit.load(Ordering::Relaxed);
+        if !is_root {
```

Comment:
Memory usage will be recorded twice.
```
Thread1MemoryTracker {
    parent_memory_tracker: Some(GlobalMemoryTracker {
        parent_memory_tracker: None
    })
}
```

Reply:
No ThreadTracker has a reference(i.e., via an `Arc`) to the global MemoryTracker.
If `ThreadTracker.mem_tracker` is None, this ThreadTracker will flush its buffer to the GLOBAL_TRACKER.

So, I think there won't be a double report.